lib/net/ssh/transport/packet_stream.rb

```diff
@@ -146,11 +146,32 @@ def enqueue_packet(payload)
 
           padding = Array.new(padding_length) { rand(256) }.pack("C*")
 
-          unencrypted_data = [packet_length, padding_length, payload, padding].pack("NCA*A*")
-          mac = client.hmac.digest([client.sequence_number, unencrypted_data].pack("NA*"))
+          if client.hmac.etm
+            debug { "using encrypt-then-mac" }
 
-          encrypted_data = client.update_cipher(unencrypted_data) << client.final_cipher
-          message = encrypted_data + mac
+            # Encrypt padding_length, payload, and padding. Take MAC
+            # from the unencrypted packet_lenght and the encrypted
+            # data.
+            length_data = [packet_length].pack("N")
```

Comment:
```suggestion
length_data = [packet_length].pack("N")
unencrypted_data = [padding_length, payload, padding].pack("CA*A*")
encrypted_data = client.update_cipher(unencrypted_data) << client.final_cipher

message = if client.hmac.etm
  # Encrypt padding_length, payload, and padding. Take MAC message = encrypted_data + mac
  # from the unencrypted packet_lenght and the encrypted data.
  mac_data = length_data + encrypted_data
  mac = client.hmac.digest([client.sequence_number, mac_data].pack("NA*"))
  mac_data + mac
else
  mac_data = length_data + unencrypted_data
  mac = client.hmac.digest([client.sequence_number, mac_data].pack("NA*"))
  encrypted_data + mac
end
```

Reply:
This change is not correct because it does not include the packet length in the encrypted data for the normal mac operations.  Fixing this would require bringing the update_cipher call into the if which is pretty much where my code was going, so I've kept my code.